ytypes/leafref.go

```diff
@@ -189,7 +189,7 @@ func dataNodesAtPath(ni *util.NodeInfo, path *gpb.Path, pathQueryNode *util.Path
 			if root.Parent == nil {
 				return nil, fmt.Errorf("no parent for leafref path at %v, with remaining path %s", ni.Schema.Path(), path)
 			}
-			if !util.IsCompressedSchema(root.Schema) && root.Parent.Schema.IsList() && util.IsValueMap(root.Parent.FieldValue) {
+			if !util.IsCompressedSchema(root.Schema) && ((root.Parent.Schema.IsList() && util.IsValueMap(root.Parent.FieldValue)) || (root.Parent.Schema.IsLeafList() && util.IsValueSlice(root.Parent.FieldValue))) {
```

Comment:
So I think it might actually be
```go
(!util.IsCompressedSchema(root.Schema) && (root.Parent.Schema.IsList() && util.IsValueMap(root.Parent.FieldValue)))
|| (root.Parent.Schema.IsLeafList() && util.IsValueSlice(root.Parent.FieldValue))
```
Since leaf-lists are never compressed.

We should also add a test for this in `TestValidateLeafRefDataCompressedSchemaListOnly` to make sure we test that the correct thing happens inside a compressed schema. I can do that in a supplemental PR if you want.

Reply:
Ack, will submit tomorrow, will take a stab at adding a test in TestValidateLeafRefDataCompressedSchemaListOnly

Reply:
I've done some testing, and it seems when compression is enabled we need a special use-case for single item lists, as they seem to be compressed into a single struct level?

Reply:
**Case1**
DataNodeAtPath got leafref with path elem:{name:".."} elem:{name:".."} elem:{name:".."} elem:{name:".."} elem:{name:"conf3"} from node path /device/root/examples/example/config/conf3-ref, field name Conf3Ref
going up data tree from type []string to *hansygot.Root_Example, schema path from parent is [config], remaining path elem:{name:".."} elem:{name:".."} elem:{name:".."} elem:{name:"conf3"}
going up data tree from type *hansygot.Root_Example to map[uint32]*hansygot.Root_Example, schema path from parent is [example], remaining path elem:{name:".."} elem:{name:".."} elem:{name:"conf3"}
going up data tree from type map[uint32]*hansygot.Root_Example to *hansygot.Root, schema path from parent is [examples], remaining path elem:{name:".."} elem:{name:"conf3"}
going up data tree from type *hansygot.Root to *hansygot.Device, schema path from parent is [root], remaining path elem:{name:"conf3"}
root element type *hansygot.Device with remaining path elem:{name:"conf3"}
. GetNode next path name:"conf3", value { { conf2 (string ptr ptr ptr), [ conf3 (string ptr ptr) ], map[42:0xc0002426c0] (map ptr ptr) } }
. getNodesContainer: schema device, next path name:"conf3", value { { conf2 (string ptr ptr ptr), [ conf3 (string ptr ptr) ], map[42:0xc0002426c0] (map ptr ptr) } }


**Case2**
DataNodeAtPath got leafref with path elem:{name:".."} elem:{name:".."} elem:{name:".."} elem:{name:".."} elem:{name:"conf3"} from node path /device/root/examples/example/config/conf3-ref, field name Conf3Ref
going up data tree from type []string to *hansygot.Root_Examples_Example, schema path from parent is [config], remaining path elem:{name:".."} elem:{name:".."} elem:{name:".."} elem:{name:"conf3"}
going up data tree from type *hansygot.Root_Examples_Example to map[uint32]*hansygot.Root_Examples_Example, schema path from parent is [example], remaining path elem:{name:".."} elem:{name:".."} elem:{name:"conf3"}
going up data tree from type map[uint32]*hansygot.Root_Examples_Example to *hansygot.Root_Examples, schema path from parent is [example], remaining path elem:{name:".."} elem:{name:"conf3"}
going up data tree from type *hansygot.Root_Examples to *hansygot.Root, schema path from parent is [examples], remaining path elem:{name:"conf3"}
root element type *hansygot.Root with remaining path elem:{name:"conf3"}
. GetNode next path name:"conf3", value { conf2 (string ptr ptr), [ conf3 (string ptr) ], { nil, map[42:0xc00027e6f0] (map ptr ptr) } }
. getNodesContainer: schema root, next path name:"conf3", value { conf2 (string ptr ptr), [ conf3 (string ptr) ], { nil, map[42:0xc00027e6f0] (map ptr ptr) } }

Reply:
Ok I think I've figured it out. I'm going to push to your branch with the fixes.

Reply:
So regarding your YANG example, compression is specific to OpenConfig, which requires every list to be wrapped underneath a container with no other sibling nodes.